scripts/update_versions.py

```diff
@@ -0,0 +1,163 @@
+import re
+import networkx as nx
+import subprocess
+import time
+
+def main():
+    # Get the paths to the go modules
+    paths = parse_go_work_paths()
+      # Create a mapping from the paths to the names of the modules
+    path_to_name = {path: parse_go_mod_name(path) for path in paths}
+    names = sorted(path_to_name.values())
+    names_to_path = {path_to_name[path]: path for path in path_to_name}
+    # build dependency graph for our commit order
+    G = build_di_graph(names, paths)
+    
+    # Perform a topological sort on the dependency graph and reverse the resulting list
+    sorted_dependencies = list(nx.topological_sort(G))[::-1]
+
+    # now we go commit every go mod file in order, replacing dependencies of prior ones.
+    update_dep_versions = {}
+    num_commits = 0
+    for dependency in sorted_dependencies:
+      # see if we need to change anything per update_dep_versions
+      mod_path = names_to_path[dependency]
+      update_dependencies(mod_path, update_dep_versions)
+      if check_for_diff_against_main(mod_path):
+        print("detected diff relative to main in " + dependency)
+        if check_for_diff_since_last_commit():
+          print("detected diff relative to last commit, committing" + dependency)
+          subprocess.run("git add .".split(" "), capture_output=True)
+          subprocess.run("git commit -m".split(" ") + ["(auto) locking dependency versions for " + dependency])
+          num_commits += 1
+        v = get_go_sum_version()
+        update_dep_versions[dependency] = str(v)
+
+    gitResetStr = "git reset HEAD" + "~"*num_commits
```

Comment:
lol. I guess this works, but funny command

Reply:
:joy: